### ISSUE TITLE
WinUI MapControl Replacement

### ISSUE THREAD

**Issue by robloo** [Mapsui/Mapsui#1125]
# Implement a WinUI Version of the Mapsui MapControl

The future of the UWP platform as a whole is questionable. However, Microsoft is continuing to support the UI layer with WinUI 3.0 which should be released and ready in November 2021. That release is missing a key control: MapControl. Mapsui is identified as one of the most promising alternatives to fill that feature gap. 

Below is a tentative list of things that need to be implemented or fixed before Mapsui seems viable in production apps. This is focused on the current UWP version of the control which seems less maintained than others. Consider this issue more of an epic and I can branch-out separate issues as needed.

## Open Issues

**High priority**

 - [ ] Make a WinUI version of the MapControl using the UWP control as a base. ~Possibly deprecate the UWP control at this point.~
 - [ ] UWP control doesn't work with Mapsui 3.0 at all. The app crashes. (potentially #978 )
 - [ ] Invalidation and reloading of tiles isn't working correctly after panning. Panning too far will show tiles in a very pixelated low resolution.
 - [ ] The Skia dependencies conflict with apps that use the latest SkiaSharp. With Mapsui 2.0.3 there are warnings such as:

   > Detected package version outside of dependency constraint: Mapsui 2.0.3 requires SkiaSharp.Views (>= 1.68.2.1 && < 1.69.0) but version SkiaSharp.Views 2.80.2 was resolved.

   This should be fixed with Mapsui v3 but the version requirements really should never be this constrained in the future. Generally, >= to a version only and not put a <= minor update.
 - [ ] ~It's possible to zoom out so far that the map goes into an error state. It appears to go outside of the viewport when zooming back in.~
       - Should be restricted in the Viewport by the application.

**Lower priority**

 - [ ] World wrapping #518. This is necessary for the user experience.
 - [ ] Prevent panning past the edges of the map when the zoom factor allows such a restriction. This ensures the map will always be 100% visible (along with world-wrapping).
 - [ ] Add a Pins collection to greatly simplify showing pins on a map

Updated as of Mapsui 2.0.3

## Additional Comments

I can put resources on fixing the high priority items. Would like to get directional feedback though and discuss how each should be addressed first.

**Comment by charlenni** [comment]
If I see it correct, then there isn't a SkiaSharp version for WinUI. So it would be difficult to add a WinUI version to MapsUI.

**Comment by robloo** [comment]
@charlenni

> If I see it correct, then there isn't a SkiaSharp version for WinUI.

There has been a lot of movement to enable this in SkiaSharp recently: https://github.com/mono/SkiaSharp/pull/1696 (merged as of yesterday) and https://github.com/mono/SkiaSharp/pull/1458. I'm not sure the status of Angle though ([if it's required](https://github.com/mono/SkiaSharp/pull/1458#issuecomment-780453574)).

Several of these updates could be done for the UWP version of the control as well. It will be 99% the same as the WInUI version anyway with only changes to namespace I am aware of. I also was probably pre-mature in my comment about deprecating the UWP version of Mapsui MapControl. It should likely stay around for years to come.

**Comment by radderz** [comment]
@robloo implementing world wrapping would be a good improvement that should be a cross platform change, you could come up with a design idea to validate first, I have implemented it in a hacky way (does work though) on my own Apps but it really would be nice to not have to be hacky. 

World wrapping: my design was I subclassed the global mercator, increased it's min/max size in the X axis to be a multiple (think I used 3x but this doesn't really need to be a specific number so long as it's a multiple) of the standard min and max, and on change of the viewport location, I would move it back to the normal equivalent position within the standard bounds if it exceeded it. I also had to update the tile fetcher to do the same, so it would return the same tile from within the bounds for extra tiles that were overs of the bounds. One issue here is the renderer for features needs to also be able to duplicate the features to the duplicated locations due to wrap that are visible within the view port.

For the viewport, that should be handled by a view port limiter, mapsui doesn't limit the viewport at all if you don't enable a limiter. 

I think getting 3.0 working correctly on UWP first is a good idea as we don't know the final APIs available within WinUI yet and the customer base able to make use of WinUI compared with WPF, ios, android etc are smaller in the short term. Microsoft has a tendency to bring out a new technology for UI that isn't quite ready for prime time in V1 but hopefully this will be a usable future UI framework.

**Comment by robloo** [comment]
@radderz Thanks for your comments

> implementing world wrapping would be a good improvement that should be a cross platform change, you could come up with a design idea to validate first

This is one of the changes I am least comfortable to make. Someone with more core experience should probably take a look and provide some solid implementation plans. Thanks for the good pointers though.

> For the viewport, that should be handled by a view port limiter, mapsui doesn't limit the viewport at all if you don't enable a limiter.

Ok, good point. I'm surprised there is no hardcoded limit to prevent the view from zooming out to an illogical level but if this is limited in the viewport only that will work.

> I think getting 3.0 working correctly on UWP first is a good idea as we don't know the final APIs available within WinUI yet

The final API's were released in 0.5. It's production ready now - especially for 3rd party controls integration like this. What is NOT ready is UWP integration and that won't come around until November with version 1.0. However, WinUI for C# and .NET Core 5/6 is ready now.

> Microsoft has a tendency to bring out a new technology for UI that isn't quite ready for prime time in V1 but hopefully this will be a usable future UI framework.

Don't get me started on this. The fact they left UWP to stagnate without updates for years now and is launching a half-baked WinUI is a big issue and is why I'm now required to switch to a different MapControl. UWP only supports .NET Standard 2.x and C# 7.3 and that's too old to use with the latest .NET libraries such as entity framework. Those that must use the latest code (also important for Uno Platform) are all going to have to migrate to WinUI quickly over the next year or so. It's going to be a messy process for sure. Regardless, 3rd party controls like this must be on the bleeding edge. Apps themselves can take longer to update.

**Comment by radderz** [comment]
@robloo I didn't realise they had finalised the APIs, that's good, I don't think it'd be that hard to create a MapControl component for WinUI assuming that the SkiaSharp libraries are supported correctly within the Framework. The main implementation hurdles per platform is really just around the touch interactions, and if they are basically the same as UWP then this shouldn't be too challenging. We have a pretty low code per platform requirement now so if you have a look at the MapControl.cs file within UWP you could try to create one of these for WinUI once SkiaSharp is ready to go. I am happy to help with the touch interactions if you get stuck as I spent a fair bit of time on it for the current platforms getting things like pinch to zoom working reliably, double tap zoom (about the point of the double tap etc). The 4 platforms that already exist will give a good guide on how to add aditional MapControl with the touch handling.

**Comment by Lee-Parker** [comment]
So SkiaSharp now has a WinUI library, I had a play around and managed to get the UWP version somewhat ported over to WinUI 3 on Desktop and thus have it tentatively working (barely tested). Essentially my changes boil down to 4 things:
- Rename all Windows.UI references to Microsoft.UI
- Update dependencies and include SkiaSharp.Views.WinUI
- Change `RunOnUIThread(Action action)` to 
`
        private void RunOnUIThread(Action action)
        {
            Task.Run(() => DispatcherQueue.TryEnqueue(Microsoft.UI.Dispatching.DispatcherQueuePriority.Normal, () => action()));
        }
`
- Change `private float GetPixelDensity()` to
`
        private float GetPixelDensity()
        {
            return (float)XamlRoot.RasterizationScale;
        }
`

The last two are because the corewindow does not exist in WinUI on Desktop (it does on WinUI on UWP)
Now I am a complete noob so I have little understanding of how to properly contribute changes. but for those who are more knowledgeable, it shouldn't take much to get an experimental branch running.

**Comment by janusw** [comment]
> So SkiaSharp now has a WinUI library

Yes, apparently mono/SkiaSharp#1298 is fixed in SkiaSharp 2.80.3. In order to make use of it in Mapsui, we need to update the SkiaSharp version first. I recently tried that in #1167, but there are errors in the tests (which are somewhat puzzling to me). Any help welcome ...

**Comment by janusw** [comment]
The update to SkiaSharp 2.80.3 is done by now. Contributions towards a MapControl for WinUI very much welcome!

**Comment by inforithmics** [comment]
Made an initial WinUI Pull request, with the changes mentioned by @Lee-Parker.

**Comment by inforithmics** [comment]
Here are the nuget packges of the pull request for feedback/testing
https://ci.appveyor.com/project/pauldendulk/mapsui/builds/40909697/artifacts

**Comment by pauldendulk** [comment]
This has been implemented in v4.
